// @microsoft/sp-build-core-tasks/lib/webpack/legacyLocalization/LegacyLocalization.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LegacyLocalization = void 0;
const path = __importStar(require("path"));
const spfx_heft_plugins_1 = require("@microsoft/spfx-heft-plugins");
const node_core_library_1 = require("@rushstack/node-core-library");
const glob_1 = __importDefault(require("glob"));
const lodash_1 = require("lodash");
const LOCALE_TOKEN = '{locale}';
const ESCAPED_TOKEN_REGEXP = new RegExp(lodash_1.escapeRegExp(lodash_1.escapeRegExp(LOCALE_TOKEN)));
const ESCAPED_TOKEN_GLOBAL_REGEXP = new RegExp(lodash_1.escapeRegExp(lodash_1.escapeRegExp(LOCALE_TOKEN)), 'g');
const LOCALE_TOKEN_REGEXP = new RegExp(LOCALE_TOKEN, 'g');
class LegacyLocalization {
    constructor(options) {
        this._legacyLocalizedExternalsByName = new Map();
        this._gulpTask = options.gulpTask;
        this._localizedResources = options.localizedResources;
        this._buildFolder = options.buildFolder;
    }
    updateWebpackConfiguration(webpackConfiguration) {
        webpackConfiguration.externals = [
            ...(webpackConfiguration.externals || []),
            ...Object.keys(this._localizedResources)
        ];
        let foundManifestPlugin = false;
        for (const plugin of webpackConfiguration.plugins) {
            if (plugin instanceof spfx_heft_plugins_1._ManifestPlugin) {
                plugin._options.tryGetLegacyLocalizedExternalByNameAsync = async (name) => {
                    return await this._tryGetLegacyLocalizedExternalsByName(name);
                };
                foundManifestPlugin = true;
            }
        }
        if (!foundManifestPlugin) {
            throw new Error('Unable to find SPFx Manifest Plugin');
        }
    }
    async _ensureLocalizedResourcesAsync() {
        if (!this._discoveredLocalizedResources) {
            const localizedResourceKeys = Object.keys(this._localizedResources);
            this._gulpTask.logVerbose(`${localizedResourceKeys.length} localized resources defined.`);
            this._discoveredLocalizedResources = new Map();
            for (const [resourceId, pattern] of Object.entries(this._localizedResources)) {
                const result = await this._collectResourcesForKeyAsync(pattern);
                this._discoveredLocalizedResources.set(resourceId, result);
            }
        }
        return this._discoveredLocalizedResources;
    }
    async _collectResourcesForKeyAsync(rawPattern) {
        // If the pattern doesn't have a leading slash, include one so the glob pattern resolves to under the root directory
        const rawPatternMatches = rawPattern.match(/(?:^\/+)(.+)/);
        const pattern = rawPatternMatches ? rawPatternMatches[1] : rawPattern;
        if (rawPatternMatches) {
            this._gulpTask.logWarning(`The localized file pattern "${rawPattern}" contains a leading slash. These patterns resolve ` +
                `to paths under the root directory ("${this._buildFolder}"), so they are intended to ` +
                'be relative. A leading slash is not recommended and may produce unexpected behavior.');
        }
        if (!pattern.match(LOCALE_TOKEN_REGEXP)) {
            // Singleton locale, no pattern token
            const fullPath = path.join(this._buildFolder, pattern);
            if (await node_core_library_1.FileSystem.existsAsync(fullPath)) {
                // Singleton found, resolve as the default locale
                return {
                    default: fullPath
                };
            }
            else {
                // Singleton not found, throw an error
                throw new Error(`Singleton resource "${rawPattern}" not found.`);
            }
        }
        else {
            const globPattern = pattern.replace(LOCALE_TOKEN_REGEXP, '*');
            const files = await node_core_library_1.LegacyAdapters.convertCallbackToPromise(glob_1.default, globPattern, {
                cwd: this._buildFolder
            });
            if (!files || files.length === 0) {
                throw new Error(`No localized files found under the root directory ("${this._buildFolder}") matching ` +
                    `pattern "${rawPattern}"`);
            }
            else {
                // This turns a pattern that looks something like this: "asdf\{locale}/{locale}/{locale}.js"
                // into a regex that looks something like this: /asdf\/(.+)\/(?:\\1)\/(?:\\1).js/
                const normalizedPattern = path.posix.resolve('/', pattern).substr(1);
                const regexPattern = new RegExp(lodash_1.escapeRegExp(normalizedPattern)
                    .replace(ESCAPED_TOKEN_REGEXP, '(.+)')
                    // this rule actually has a bug, so we'll remove this exception when it gets fixed. VSO#513580
                    // tslint:disable-next-line:no-octal-literal
                    .replace(ESCAPED_TOKEN_GLOBAL_REGEXP, '(?:\\1)'));
                const result = {};
                for (const file of files) {
                    const posixFile = file.replace(/\\/g, '/');
                    const match = posixFile.match(regexPattern);
                    if (!match) {
                        this._gulpTask.logWarning(`Unexpected file path: "${file}"`);
                    }
                    else {
                        const locale = match[1];
                        this._gulpTask.logVerbose(`Found file "${file}", identified as locale "${locale}"`);
                        result[locale] = path.join(this._buildFolder, file);
                    }
                }
                return result;
            }
        }
    }
    async _tryGetLegacyLocalizedExternalsByName(externalName) {
        let result = this._legacyLocalizedExternalsByName.get(externalName);
        if (!result) {
            const discoveredLocalizedResources = await this._ensureLocalizedResourcesAsync();
            const localizedFiles = discoveredLocalizedResources.get(externalName);
            if (localizedFiles) {
                const manifestModulePaths = {};
                const assets = {};
                for (const [localeName, filePath] of Object.entries(localizedFiles)) {
                    const fileContents = await node_core_library_1.FileSystem.readFileAsync(filePath);
                    const fileHash = spfx_heft_plugins_1._getFullHash(fileContents);
                    const assetFilename = `${externalName}_${localeName}_${fileHash}${path.extname(filePath)}`;
                    manifestModulePaths[spfx_heft_plugins_1._normalizeLocaleName(localeName)] = assetFilename;
                    assets[assetFilename] = fileContents;
                }
                const manifestModule = {
                    type: 'localizedPath',
                    paths: manifestModulePaths,
                    defaultPath: spfx_heft_plugins_1._LocalizedManifestModuleProcessor.findDefaultString(manifestModulePaths)
                };
                result = {
                    moduleConfiguration: manifestModule,
                    assets: assets
                };
                this._legacyLocalizedExternalsByName.set(externalName, result);
            }
        }
        return result;
    }
}
exports.LegacyLocalization = LegacyLocalization;
